packages/core/src/common/reference.spec.ts

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (C) 2018 TypeFox and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import * as assert from "assert";
+import { ReferenceCollection } from './reference';
+import { Disposable } from "./disposable";
+
+describe('reference', () => {
+
+    it('dispose a single reference', async () => {
+        const expectation: { disposed: boolean } = { disposed: false };
+        const references = new ReferenceCollection<string, Disposable>(key => ({
+            key, dispose: () => {
+                expectation.disposed = true;
+            }
+        }));
+        assert.ok(!references.has("a"));
+        assert.ok(!expectation.disposed);
+
+        const reference = await references.acquire("a");
+        assert.ok(references.has("a"));
+        assert.ok(!expectation.disposed);
+
+        reference.dispose();
+        assert.ok(!references.has("a"));
+        assert.ok(expectation.disposed);
+    });
+
+    it('dispose 2 references', async () => {
+        const expectation: { disposed: boolean } = { disposed: false };
+        const references = new ReferenceCollection<string, Disposable>(key => ({
+            key, dispose: () => {
+                expectation.disposed = true;
+            }
+        }));
+        assert.ok(!references.has("a"));
+        assert.ok(!expectation.disposed);
+
+        const reference = await references.acquire("a");
+        const reference2 = await references.acquire("a");
+        assert.ok(references.has("a"));
+        assert.ok(!expectation.disposed);
+
+        reference.dispose();
+        assert.ok(references.has("a"));
+        assert.ok(!expectation.disposed);
+
+        reference2.dispose();
+        assert.ok(!references.has("a"));
+        assert.ok(expectation.disposed);
+    });
+
+    it('dispose an object with 2 references', async () => {
+        const expectation: { disposed: boolean } = { disposed: false };
+        const references = new ReferenceCollection<string, Disposable>(key => ({
+            key, dispose: () => {
+                expectation.disposed = true;
+            }
+        }));
+        assert.ok(!references.has("a"));
+        assert.ok(!expectation.disposed);
+
+        await references.acquire("a");
+        const reference = await references.acquire("a");
+        assert.ok(references.has("a"));
+        assert.ok(!expectation.disposed);
+
+        reference.object.dispose();
+        assert.ok(!references.has("a"));
+        assert.ok(expectation.disposed);
+    });
```

Comment:
So if two different call `acquire` they'll get the same reference is that it? And if one of them disposes it what happens in that case? Looking like a great change so far!

Reply:
Each call a new reference to the same object. If all references are disposed then an object gets disposed.

Reply:
Okay perfect :)